File: code/ex338.java
class Solution {
    public int[] countBits(int num) {
        int[] dp = new int[num + 1];
        dp[0] = 0;    if (num == 0)    return dp;
        
        
        int i = 1;
        int exp = 1;
        
        while (i <= num) {
            int gap = (int)Math.pow(2, exp - 1);
            while (i < gap * 2 && i <= num) {
                dp[i] = dp[i - gap] + 1;
                i++;
            }
            exp++;
        }

        return dp;
    }
}

File: code/ex338.cpp
class Solution {
public:
    vector<int> countBits(int num) {
        vector<int> dp(num + 1);
        int exp = 0;
        
        for (int i = 1; i <= num; i++) {
            if (i >= pow(2, exp + 1))    exp++;
            dp[i] = dp[i - pow(2, exp)] + 1;
        }
        
        return dp;
    }
};